File: backend/app/schemas/ai/ai_tool.py
"""Tool / ToolGroup 管理 Schemas（字段命名对齐前端测试场景，使用 camelCase）。

字段清单（来自 docs/ai-scene.md 测试场景 1-11）：
toolKey / displayName / category / type / className / methodName /
description / configSchema / configValue / inputSchema / outputSchema /
status / isSystem / sort
"""
from __future__ import annotations
import re
from typing import Any, Dict, List, Optional
from datetime import datetime

from pydantic import BaseModel, Field, field_validator

TOOL_KEY_RE = re.compile(r"^[A-Za-z][A-Za-z0-9_]*$")
VALID_TYPES = {"custom", "skill", "mcp", "group", "sqlbot", "agentscope_builtin", "custom_dev"}
VALID_STATUS = {"enabled", "disabled"}


class AiToolCreate(BaseModel):
    """创建工具"""
    toolKey: str = Field(..., min_length=1, max_length=100, description="工具标识")
    displayName: str = Field(..., min_length=1, max_length=200, description="显示名称")
    category: Optional[str] = Field(None, max_length=100, description="分类")
    type: str = Field("custom", description="工具类型 custom/skill/mcp/group")
    className: Optional[str] = Field(None, max_length=500, description="实现类全限定名")
    methodName: Optional[str] = Field(None, max_length=200, description="方法名")
    description: Optional[str] = Field(None, max_length=2000)
    configSchema: Optional[Dict[str, Any]] = Field(default=None, description="配置 Schema")
    configValue: Optional[Dict[str, Any]] = Field(default=None, description="配置值")
    inputSchema: Optional[Dict[str, Any]] = Field(default=None, description="输入参数 JSON Schema")
    outputSchema: Optional[Dict[str, Any]] = Field(default=None, description="输出参数 JSON Schema")
    status: str = Field("enabled", description="状态 enabled/disabled")
    isSystem: bool = Field(False, description="是否系统内置")
    sort: int = Field(0, description="排序")

    @field_validator("toolKey")
    @classmethod
    def _check_tool_key(cls, v: str) -> str:
        if not TOOL_KEY_RE.match(v):
            raise ValueError("toolKey 只能包含字母、数字、下划线，且以字母开头")
        return v

    @field_validator("type")
    @classmethod
    def _check_type(cls, v: str) -> str:
        if v not in VALID_TYPES:
            raise ValueError(f"type 必须是 {sorted(VALID_TYPES)} 之一")
        return v

    @field_validator("status")
    @classmethod
    def _check_status(cls, v: str) -> str:
        if v not in VALID_STATUS:
            raise ValueError(f"status 必须是 {sorted(VALID_STATUS)} 之一")
        return v


class AiToolUpdate(BaseModel):
    """更新工具（全字段可选，id 必填）"""
    id: int = Field(..., description="主键 ID")
    toolKey: Optional[str] = Field(None, min_length=1, max_length=100)
    displayName: Optional[str] = Field(None, min_length=1, max_length=200)
    category: Optional[str] = Field(None, max_length=100)
    type: Optional[str] = Field(None)
    className: Optional[str] = Field(None, max_length=500)
    methodName: Optional[str] = Field(None, max_length=200)
    description: Optional[str] = Field(None, max_length=2000)
    configSchema: Optional[Dict[str, Any]] = None
    configValue: Optional[Dict[str, Any]] = None
    inputSchema: Optional[Dict[str, Any]] = None
    outputSchema: Optional[Dict[str, Any]] = None
    status: Optional[str] = None
    isSystem: Optional[bool] = None
    sort: Optional[int] = None

    @field_validator("toolKey")
    @classmethod
    def _check_tool_key(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and not TOOL_KEY_RE.match(v):
            raise ValueError("toolKey 只能包含字母、数字、下划线，且以字母开头")
        return v

    @field_validator("type")
    @classmethod
    def _check_type(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in VALID_TYPES:
            raise ValueError(f"type 必须是 {sorted(VALID_TYPES)} 之一")
        return v

    @field_validator("status")
    @classmethod
    def _check_status(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in VALID_STATUS:
            raise ValueError(f"status 必须是 {sorted(VALID_STATUS)} 之一")
        return v


class AiToolResp(BaseModel):
    """工具响应"""
    id: int
    toolKey: str
    displayName: str
    category: Optional[str] = None
    type: str
    className: Optional[str] = None
    methodName: Optional[str] = None
    description: Optional[str] = None
    configSchema: Optional[Dict[str, Any]] = None
    configValue: Optional[Dict[str, Any]] = None
    inputSchema: Optional[Dict[str, Any]] = None
    outputSchema: Optional[Dict[str, Any]] = None
    status: str
    isSystem: bool
    sort: int
    creator: Optional[str] = None
    updater: Optional[str] = None
    createdAt: Optional[datetime] = None
    updatedAt: Optional[datetime] = None


class AiToolSimpleResp(BaseModel):
    """工具简易响应（下拉选择用）"""
    id: int
    toolKey: str
    displayName: str
    category: Optional[str] = None


# ============= ToolGroup =============

class AiToolGroupCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    displayName: Optional[str] = Field(None, max_length=200)
    description: Optional[str] = Field(None, max_length=2000)
    instructions: Optional[str] = Field(None, max_length=4000)
    isActive: bool = Field(False)
    sort: int = Field(0)


class AiToolGroupUpdate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    displayName: Optional[str] = Field(None, max_length=200)
    description: Optional[str] = Field(None, max_length=2000)
    instructions: Optional[str] = Field(None, max_length=4000)
    isActive: Optional[bool] = None
    sort: Optional[int] = None


class AiToolGroupResp(BaseModel):
    id: int
    name: str
    displayName: Optional[str] = None
    description: Optional[str] = None
    instructions: Optional[str] = None
    isActive: bool
    sort: int
    toolCount: int = 0
    createdAt: Optional[datetime] = None
    updatedAt: Optional[datetime] = None


class AiToolGroupMemberAdd(BaseModel):
    toolKeys: List[str] = Field(..., min_length=1, description="要加入分组的 toolKey 列表")


class AiToolTestRequest(BaseModel):
    """工具测试请求"""
    toolKey: Optional[str] = Field(None, description="按 toolKey 测试（优先）")
    id: Optional[int] = Field(None, description="按 id 测试")
    inputs: Dict[str, Any] = Field(default_factory=dict, description="测试输入参数")
    className: Optional[str] = None
    methodName: Optional[str] = None

File: frontend/src/api/ai-tool.ts
/**
 * AI 工具管理 API
 */
import request from '@/utils/request'

// 工具类型
export const TOOL_TYPES = [
  { value: 'custom', label: '自定义' },
  { value: 'skill', label: '技能' },
  { value: 'mcp', label: 'MCP' },
  { value: 'group', label: '分组' },
  { value: 'sqlbot', label: 'SQLBot' },
  { value: 'agentscope_builtin', label: 'AgentScope内置' },
  { value: 'custom_dev', label: '自定义开发' }
] as const

// 默认分类（与后端 /categories 兜底一致）
export const DEFAULT_CATEGORIES = [
  '信息查询', '数据处理', 'AI增强',
  // AgentScope 内置工具分类
  '文件系统', '任务管理', 'Shell执行',
  // 项目自定义开发工具分类
  '文档处理', '自动化', '消息通知', '数据分析',
] as const

export interface AiTool {
  id: number
  toolKey: string
  displayName: string
  category?: string | null
  type: string
  className?: string | null
  methodName?: string | null
  description?: string | null
  configSchema?: Record<string, any> | null
  configValue?: Record<string, any> | null
  inputSchema?: Record<string, any> | null
  outputSchema?: Record<string, any> | null
  status: string
  isSystem: boolean
  sort: number
  creator?: string | null
  updater?: string | null
  createdAt?: string
  updatedAt?: string
}

export interface ToolPageResp {
  data: AiTool[]
  total: number
  page: number
  pageSize: number
}

export interface ToolSimpleItem {
  id: number
  toolKey: string
  displayName: string
  category?: string | null
}

export interface ToolGroup {
  id: number
  name: string
  displayName?: string | null
  description?: string | null
  instructions?: string | null
  isActive: boolean
  sort: number
  toolCount: number
  createdAt?: string
  updatedAt?: string
}

export interface ToolGroupPageResp {
  data: ToolGroup[]
  total: number
  page: number
  pageSize: number
}

// ========== 工具 CRUD ==========

export function getToolPage(params: {
  toolKey?: string
  displayName?: string
  category?: string
  type?: string
  status?: string
  isSystem?: boolean
  page?: number
  pageSize?: number
}) {
  return request.get<ToolPageResp>('/api/v1/admin/ai/tool/page', { params })
}

export function getToolDetail(id: number) {
  return request.get<AiTool>(`/api/v1/admin/ai/tool/get`, { params: { id } })
}

export function createTool(data: Partial<AiTool>) {
  return request.post<{ id: number; message: string }>('/api/v1/admin/ai/tool/create', data)
}

export function updateTool(data: Partial<AiTool> & { id: number }) {
  return request.post<{ id: number; message: string }>('/api/v1/admin/ai/tool/update', data)
}

export function deleteTool(id: number) {
  return request.delete<{ message: string }>('/api/v1/admin/ai/tool/delete', { params: { id } })
}

export function deleteToolList(ids: number[]) {
  return request.delete<{ message: string }>('/api/v1/admin/ai/tool/delete-list', {
    params: { ids: ids.join(',') }
  })
}

export function getToolSimpleList(params?: { category?: string }) {
  return request.get<ToolSimpleItem[]>('/api/v1/admin/ai/tool/simple-list', { params })
}

export function testTool(payload: {
  toolKey?: string
  id?: number
  inputs?: Record<string, any>
  className?: string
  methodName?: string
}) {
  return request.post<{
    success: boolean
    output: any
    error: string | null
    target: string
  }>('/api/v1/admin/ai/tool/test', payload)
}

export function refreshToolCache() {
  return request.post<{ message: string; count: number }>('/api/v1/admin/ai/tool/refresh-cache')
}

export function getToolCategories() {
  return request.get<string[]>('/api/v1/admin/ai/tool/categories')
}

// ========== 工具分组 ==========

export function getToolGroupPage(params: { name?: string; page?: number; pageSize?: number }) {
  return request.get<ToolGroupPageResp>('/api/v1/admin/ai/tool/group/page', { params })
}

export function createToolGroup(data: {
  name: string
  displayName?: string
  description?: string
  instructions?: string
  isActive?: boolean
  sort?: number
}) {
  return request.post<{ id: number; message: string }>('/api/v1/admin/ai/tool/group/create', data)
}

export function updateToolGroup(data: {
  id: number
  name: string
  displayName?: string
  description?: string
  instructions?: string
  isActive?: boolean
  sort?: number
}) {
  return request.post<{ id: number; message: string }>('/api/v1/admin/ai/tool/group/update', data)
}

export function deleteToolGroup(id: number) {
  return request.delete<{ message: string }>('/api/v1/admin/ai/tool/group/delete', { params: { id } })
}

export function addGroupMembers(groupId: number, toolKeys: string[]) {
  return request.post<{ message: string }>('/api/v1/admin/ai/tool/group/members/add', {
    toolKeys
  }, { params: { group_id: groupId } })
}

export function removeGroupMember(groupId: number, toolKey: string) {
  return request.post<{ message: string }>('/api/v1/admin/ai/tool/group/members/remove', null, {
    params: { group_id: groupId, toolKey }
  })
}

export function getGroupMembers(groupId: number) {
  return request.get<AiTool[]>('/api/v1/admin/ai/tool/group/members', { params: { group_id: groupId } })
}

// ========== 工具启用/禁用 ==========

export function enableTool(id: number) {
  return request.post<{ id: number; status: string; message: string }>(
    '/api/v1/admin/ai/tool/enable',
    null,
    { params: { id } }
  )
}

export function disableTool(id: number) {
  return request.post<{ id: number; status: string; message: string }>(
    '/api/v1/admin/ai/tool/disable',
    null,
    { params: { id } }
  )
}

// ========== SQLBot 数据源管理 ==========

export const SQLBOT_TOOL_TYPE = 'sqlbot'

export interface SqlBotDatasource {
  id: number
  name: string
  dbType: string
  database: string
  description: string
  tableCount: number
}

export interface SqlBotTableColumn {
  name: string
  type: string
  comment?: string
  nullable?: boolean
  isSensitive?: boolean
  suggestedMask?: string
}

export interface SqlBotTableInfo {
  id?: number
  datasourceId: number
  datasourceName?: string | null
  tableName: string
  tableComment?: string | null
  businessLine?: string | null
  columns?: SqlBotTableColumn[]
  source: string
  confirmed: boolean
  status: string
  syncedAt?: string | null
}

export interface DdlImportDetail {
  tableName: string
  columnCount: number
  sensitiveColumns: string[]
  created: number
  updated: number
}

export interface DdlImportResult {
  datasourceId: number
  imported: number
  skipped: number
  errors: string[]
  details: DdlImportDetail[]
}

export function getSqlBotDatasources() {
  return request.get<{ datasources: SqlBotDatasource[]; total: number }>(
    '/api/v1/admin/ai/sqlbot-tool/datasource/list'
  )
}

export function getSqlBotTables(datasourceId: number) {
  return request.get<SqlBotTableInfo[]>(`/api/v1/admin/ai/sqlbot-tool/tables`, {
    params: { datasourceId }
  })
}

export function importSqlBotDdl(payload: { datasourceId: number; datasourceName?: string; ddlText: string }) {
  return request.post<DdlImportResult>('/api/v1/admin/ai/sqlbot-tool/tables/import-ddl', payload)
}

export function updateSqlBotTableColumns(
  id: number,
  payload: {
    tableComment?: string
    businessLine?: string
    columns?: SqlBotTableColumn[]
    confirmed?: boolean
    status?: string
    sort?: number
  }
) {
  return request.post<SqlBotTableInfo>(
    `/api/v1/admin/ai/sqlbot-tool/tables/${id}/columns-update`,
    payload
  )
}

export function deleteSqlBotTable(id: number) {
  return request.delete<{ message: string }>(`/api/v1/admin/ai/sqlbot-tool/tables`, {
    params: { id }
  })
}

// ========== SQLBot 工具配置校验/试跑/发布 ==========

export interface ConfigValidateResp {
  valid: boolean
  errors: string[]
  warnings: string[]
  inputSchema?: Record<string, any>
  renderedQuestionPreview?: string
}

export function validateSqlBotConfig(configValue: Record<string, any>) {
  return request.post<ConfigValidateResp>('/api/v1/admin/ai/sqlbot-tool/config/validate', {
    configValue
  })
}

export function testSqlBotTool(payload: {
  toolKey?: string
  configValue?: Record<string, any>
  inputs?: Record<string, any>
}) {
  return request.post<{ success: boolean; output: any; warnings: string[]; target: string }>(
    '/api/v1/admin/ai/sqlbot-tool/test',
    payload
  )
}

export function publishSqlBotTool(data: {
  toolKey: string
  displayName: string
  category?: string
  description?: string
  configValue: Record<string, any>
  status?: string
  sort?: number
  groupName?: string
}) {
  return request.post<{ id: number; message: string }>('/api/v1/admin/ai/sqlbot-tool/publish', data)
}

export function getSqlBotConfigSchema() {
  return request.get<{
    configVersion: string
    classPath: string
    schema: Record<string, any>
  }>('/api/v1/admin/ai/sqlbot-tool/config-schema')
}

